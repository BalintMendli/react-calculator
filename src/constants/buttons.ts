const buttons = [
  { label: 'AC', id: 'clear', type: 'clear', position: { x: 6, y: 1 } },
  { label: 'C', id: 'clear', type: 'clear', position: { x: 5, y: 1 } },
  { label: '/', id: 'divide', type: 'operation', position: { x: 3, y: 1 } },
  { label: '*', id: 'multiply', type: 'operation', position: { x: 4, y: 1 } },
  { label: '7', id: 'seven', type: 'number', position: { x: 3, y: 2 } },
  { label: '8', id: 'eight', type: 'number', position: { x: 4, y: 2 } },
  { label: '9', id: 'nine', type: 'number', position: { x: 5, y: 2 } },
  { label: '-', id: 'subtract', type: 'operation', position: { x: 6, y: 2 } },
  { label: '4', id: 'four', type: 'number', position: { x: 3, y: 3 } },
  { label: '5', id: 'five', type: 'number', position: { x: 4, y: 3 } },
  { label: '6', id: 'six', type: 'number', position: { x: 5, y: 3 } },
  { label: '+', id: 'add', type: 'operation', position: { x: 6, y: 3 } },
  { label: '1', id: 'one', type: 'number', position: { x: 3, y: 4 } },
  { label: '2', id: 'two', type: 'number', position: { x: 4, y: 4 } },
  { label: '3', id: 'three', type: 'number', position: { x: 5, y: 4 } },
  { label: '=', id: 'equals', type: 'equals', position: { x: 6, y: [4, 6] } },
  { label: '0', id: 'zero', type: 'number', position: { x: [3, 5], y: 5 } },
  { label: '.', id: 'decimal', type: 'decimal', position: { x: 5, y: 5 } },
  { label: '(', id: 'open-paren', type: 'paren', position: { x: 1, y: 1 } },
  { label: ')', id: 'close-paren', type: 'paren', position: { x: 2, y: 1 } },
  { label: 'sin', id: 'sin', type: 'ext', position: { x: 1, y: 2 } },
  { label: 'cos', id: 'cos', type: 'ext', position: { x: 1, y: 3 } },
  { label: 'tan', id: 'tan', type: 'ext', position: { x: 1, y: 4 } },
  { label: 'π', id: 'tan', type: 'ext', position: { x: 1, y: 5 } },
  { label: '√', id: 'tan', type: 'ext', position: { x: 2, y: 2 } },
  { label: 'x²', id: 'tan', type: 'ext', position: { x: 2, y: 3 } },
  { label: 'log', id: 'tan', type: 'ext', position: { x: 2, y: 4 } },
  { label: 'x!', id: 'tan', type: 'ext', position: { x: 2, y: 5 } },
] as const;

export type Button = typeof buttons[number];

export type Label = typeof buttons[number]['label'];

export type Id = typeof buttons[number]['id'];

export type BtnType = typeof buttons[number]['type'];

export type Pos = typeof buttons[number]['position'];

export default buttons;
