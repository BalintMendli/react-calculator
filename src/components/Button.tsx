import React from 'react';
import styled from 'styled-components/macro';
import { Label, Id, Pos } from 'constants/buttons';
import { Color } from 'constants/colors';
import { getGridPos } from 'logic/utils';

interface ButtonProps {
  label: Label;
  id: Id;
  pos: Pos;
  color: Color;
  handleClick: (label: Label) => void;
}

interface StyledProps {
  color: Color;
  gridPos: [string, string];
}

const StyledButton = styled.button<StyledProps>`
  grid-column: ${({ gridPos }): string => gridPos[0]};
  grid-row: ${({ gridPos }): string => gridPos[1]};
  background: ${({ color }): string => color};
  padding: 0;
  font: inherit;
  font-size: 1.1rem;
  outline: inherit;
  color: #fff;
  display: flex;
  align-items: center;
  justify-content: center;
  border: 1px solid #000;
  border-radius: 5px;
  cursor: pointer;
  user-select: none;
  opacity: 0.85;
  transition: opacity 250ms ease-in-out;

  :hover {
    opacity: 1;
  }
`;

const Button = ({
  label,
  id,
  handleClick,
  pos,
  color,
}: ButtonProps): JSX.Element => {
  const handleBtnClick = (): void => {
    handleClick(label);
  };

  return (
    <StyledButton
      type="button"
      id={id}
      color={color}
      gridPos={getGridPos(pos)}
      onClick={handleBtnClick}
    >
      {label}
    </StyledButton>
  );
};

export default Button;
