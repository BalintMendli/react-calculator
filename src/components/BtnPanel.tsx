import React from 'react';
import styled from 'styled-components/macro';
import buttons, { Label } from 'constants/buttons';
import { getColor } from 'constants/colors';
import { rows, columns } from 'constants/dimensions';
import { getDimStyle } from 'logic/utils';
import Button from './Button';

interface BtnPanelProps {
  handleClick: (label: Label) => void;
}

interface StyledProps {
  width: number;
  height: number;
}

const StyledBtnPanel = styled.div<StyledProps>`
  background-color: #0b0c11;
  flex: 1;
  display: grid;
  grid-template-columns: ${(props): string => getDimStyle(props.width)};
  grid-template-rows: ${(props): string => getDimStyle(props.height)};
  box-shadow: 1px 2px 10px rgba(0, 0, 0, 0.5);
  border-radius: 5px;
`;

const BtnPanel = ({ handleClick }: BtnPanelProps): JSX.Element => (
  <StyledBtnPanel id="btn-panel" width={columns} height={rows}>
    {buttons.map(btn => (
      <Button
        key={btn.label}
        label={btn.label}
        id={btn.id}
        pos={btn.position}
        color={getColor(btn)}
        handleClick={handleClick}
      />
    ))}
  </StyledBtnPanel>
);

export default BtnPanel;
