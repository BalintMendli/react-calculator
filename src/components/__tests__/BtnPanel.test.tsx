import React from 'react';
import BtnPanel from 'components/BtnPanel';
import Button from 'components/Button';
import buttons from 'constants/buttons';
import { shallow, ShallowWrapper } from 'enzyme';

describe('BtnPanel component', () => {
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    wrapper = shallow(<BtnPanel handleClick={jest.fn()} />);
  });

  it('renders correctly', () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });

  it('renders btn-panel div', () => {
    expect(wrapper.find('#btn-panel').length).toEqual(1);
  });

  it('renders Button components', () => {
    expect(wrapper.find(Button).length).toEqual(buttons.length);
  });
});
