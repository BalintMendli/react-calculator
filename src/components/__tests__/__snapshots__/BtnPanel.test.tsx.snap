// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BtnPanel component renders correctly 1`] = `
"<BtnPanel__StyledBtnPanel id=\\"btn-panel\\" width={6} height={5}>
  <Button label=\\"AC\\" id=\\"clear\\" pos={{...}} color=\\"#a5243d\\" handleClick={[Function: mockConstructor]} />
  <Button label=\\"C\\" id=\\"clear\\" pos={{...}} color=\\"#a5243d\\" handleClick={[Function: mockConstructor]} />
  <Button label=\\"/\\" id=\\"divide\\" pos={{...}} color=\\"#798086\\" handleClick={[Function: mockConstructor]} />
  <Button label=\\"*\\" id=\\"multiply\\" pos={{...}} color=\\"#798086\\" handleClick={[Function: mockConstructor]} />
  <Button label=\\"7\\" id=\\"seven\\" pos={{...}} color=\\"#42464a\\" handleClick={[Function: mockConstructor]} />
  <Button label=\\"8\\" id=\\"eight\\" pos={{...}} color=\\"#42464a\\" handleClick={[Function: mockConstructor]} />
  <Button label=\\"9\\" id=\\"nine\\" pos={{...}} color=\\"#42464a\\" handleClick={[Function: mockConstructor]} />
  <Button label=\\"-\\" id=\\"subtract\\" pos={{...}} color=\\"#798086\\" handleClick={[Function: mockConstructor]} />
  <Button label=\\"4\\" id=\\"four\\" pos={{...}} color=\\"#42464a\\" handleClick={[Function: mockConstructor]} />
  <Button label=\\"5\\" id=\\"five\\" pos={{...}} color=\\"#42464a\\" handleClick={[Function: mockConstructor]} />
  <Button label=\\"6\\" id=\\"six\\" pos={{...}} color=\\"#42464a\\" handleClick={[Function: mockConstructor]} />
  <Button label=\\"+\\" id=\\"add\\" pos={{...}} color=\\"#798086\\" handleClick={[Function: mockConstructor]} />
  <Button label=\\"1\\" id=\\"one\\" pos={{...}} color=\\"#42464a\\" handleClick={[Function: mockConstructor]} />
  <Button label=\\"2\\" id=\\"two\\" pos={{...}} color=\\"#42464a\\" handleClick={[Function: mockConstructor]} />
  <Button label=\\"3\\" id=\\"three\\" pos={{...}} color=\\"#42464a\\" handleClick={[Function: mockConstructor]} />
  <Button label=\\"=\\" id=\\"equals\\" pos={{...}} color=\\"#466995\\" handleClick={[Function: mockConstructor]} />
  <Button label=\\"0\\" id=\\"zero\\" pos={{...}} color=\\"#42464a\\" handleClick={[Function: mockConstructor]} />
  <Button label=\\".\\" id=\\"decimal\\" pos={{...}} color=\\"#42464a\\" handleClick={[Function: mockConstructor]} />
  <Button label=\\"(\\" id=\\"open-paren\\" pos={{...}} color=\\"#595a5b\\" handleClick={[Function: mockConstructor]} />
  <Button label=\\")\\" id=\\"close-paren\\" pos={{...}} color=\\"#595a5b\\" handleClick={[Function: mockConstructor]} />
  <Button label=\\"sin\\" id=\\"sin\\" pos={{...}} color=\\"#595a5b\\" handleClick={[Function: mockConstructor]} />
  <Button label=\\"cos\\" id=\\"cos\\" pos={{...}} color=\\"#595a5b\\" handleClick={[Function: mockConstructor]} />
  <Button label=\\"tan\\" id=\\"tan\\" pos={{...}} color=\\"#595a5b\\" handleClick={[Function: mockConstructor]} />
  <Button label=\\"π\\" id=\\"tan\\" pos={{...}} color=\\"#595a5b\\" handleClick={[Function: mockConstructor]} />
  <Button label=\\"√\\" id=\\"tan\\" pos={{...}} color=\\"#595a5b\\" handleClick={[Function: mockConstructor]} />
  <Button label=\\"x²\\" id=\\"tan\\" pos={{...}} color=\\"#595a5b\\" handleClick={[Function: mockConstructor]} />
  <Button label=\\"log\\" id=\\"tan\\" pos={{...}} color=\\"#595a5b\\" handleClick={[Function: mockConstructor]} />
  <Button label=\\"x!\\" id=\\"tan\\" pos={{...}} color=\\"#595a5b\\" handleClick={[Function: mockConstructor]} />
</BtnPanel__StyledBtnPanel>"
`;
